src/ast/nodes/shared/ClassNode.ts

```diff
@@ -1,47 +1,242 @@
+			}
+			return true;
+		} else if (typeof key === 'string' && path.length === 1) {
+			return mayHaveGetterSetterEffect(this.body, 'get', true, key, context);
+		} else {
+			return true;
+		}
 	}
 
-	hasEffectsWhenAssignedAtPath(path: ObjectPath) {
-		if (path.length <= 1) return false;
-		return path.length > 2 || path[0] !== 'prototype';
+	hasEffectsWhenAssignedAtPath(path: ObjectPath, context: HasEffectsContext) {
+		if (this.deoptimizedStatic) return true;
+		if (this.superClass && this.superClass.hasEffectsWhenAssignedAtPath(path, context)) return true;
+		const key = path[0];
+		if (key === 'prototype') {
+			if (path.length === 1) return false;
+			if (this.deoptimizedPrototype) return true;
+			const key2 = path[1];
+			if (path.length === 2 && typeof key2 === 'string') {
+				return mayHaveGetterSetterEffect(this.body, 'set', false, key2, context);
+			}
+			return true;
+		} else if (typeof key === 'string' && path.length === 1) {
+			return mayHaveGetterSetterEffect(this.body, 'set', true, key, context);
+		} else {
+			return true;
+		}
 	}
 
 	hasEffectsWhenCalledAtPath(
 		path: ObjectPath,
 		callOptions: CallOptions,
 		context: HasEffectsContext
 	) {
-		if (!callOptions.withNew) return true;
-		return (
-			this.body.hasEffectsWhenCalledAtPath(path, callOptions, context) ||
-			(this.superClass !== null &&
-				this.superClass.hasEffectsWhenCalledAtPath(path, callOptions, context))
-		);
+		if (callOptions.withNew) {
+			return path.length > 0 ||
+				(this.classConstructor !== null &&
+				 this.classConstructor.hasEffectsWhenCalledAtPath(EMPTY_PATH, callOptions, context)) ||
+				(this.superClass !== null &&
+				 this.superClass.hasEffectsWhenCalledAtPath(path, callOptions, context));
+		} else {
+			if (path.length !== 1 || this.deoptimizedStatic) return true;
+			const key = path[0];
+			const definition = typeof key === 'string' && this.getStaticPropertyMap()[key];
+			if (!definition) return true;
+			return definition.hasEffectsWhenCalledAtPath([], callOptions, context);
+		}
 	}
 
 	initialise() {
 		if (this.id !== null) {
 			this.id.declare('class', this);
 		}
+		for (const method of this.body.body) {
+			if (method instanceof MethodDefinition && method.kind === 'constructor') {
+				this.classConstructor = method;
+				return;
+			}
+		}
+		this.classConstructor = null;
+	}
+
+	mayModifyThisWhenCalledAtPath(
+		path: ObjectPath
+	) {
+		const key = path[0];
+		const definition = typeof key === 'string' && this.getStaticPropertyMap()[key];
+		if (!definition || this.deoptimizedStatic) return true;
+		return definition.mayModifyThisWhenCalledAtPath(path.slice(1));
+	}
+
+	private getStaticPropertyMap(): {[name: string]: ExpressionNode} {
+		if (this.staticPropertyMap) return this.staticPropertyMap;
+
+		const propertyMap = this.staticPropertyMap = Object.create(null);
+		const seen: {[name: string]: boolean} = Object.create(null);
+		for (const definition of this.body.body) {
```

Comment:
Can this be done by building an up-front lookup table similar to how we do this for object expression? For classes with many definitions, this will scale badly as `mayModifyThisWhenCalledAtPath` is called during tree-shaking with potentially many passes (dozens in larger code bases) with potentially many calls per pass. This is a highly performance critical code path.
Maybe some code can be even extracted from/shared with object expression. The object expression logic is indeed very powerful as it will also pre-resolve computed properties if possible, handle quoted properties, handle multiple definitions of the same name and handle unknown computed properties that might overwrite previous properties.

Reply:
I decided against sharing code with ObjectExpression because there's a lot of small differences between how this works and how objects work (in syntax tree shape, handling of builtin properties, static/prototype distinction, and so on) and the helper functions would become monsters with a list of boolean parameters, and I prefer some vague duplication over that most of the time.

Building up a table is probably a good idea. Looking into that.

Reply:
Wait, the thing you're commenting on here _is_ building up a table. The suggestions holds for `mayHaveGetterSetterEffect`, but, as far as I can see, only there.